File: Sudoku/sudoku.py
class Sudoku:
    puzzle = None

    def __init__(self, file_name):
        self.puzzle = self.read_file(file_name)

    def read_file(self, file_name):
        file = open(file_name, 'r')
        lines = file.readlines()
        grid = []

        for line in lines:
            row = []
            for num in line.split(' '):
                row.append(int(num))
            grid.append(row)
        return grid

    def remove_row_values(self, arr, x):
        # remove elements that appear in the row
        for i in range(9):
            num = self.puzzle[x][i]
            if num in arr:
                arr.remove(num)

    def remove_column_values(self, arr, y):
        # remove elements that appear in the column
        for i in range(9):
            num = self.puzzle[i][y]
            if num in arr:
                arr.remove(num)

    def remove_box_values(self, arr, x, y):
        # remove elements that appear in the 3x3
        x_start, x_end = self.box_range(x)
        y_start, y_end = self.box_range(y)

        for i in range(x_start, x_end):
            for j in range(y_start, y_end):
                num = self.puzzle[i][j]
                if num in arr:
                    arr.remove(num)

    def box_range(self, num):
        # return the start and stop for box
        for i in range(4):
            if i * 3 <= num < (i + 1) * 3:
                return i * 3, (i + 1) * 3

    def print_grid(self):
        for row in self.puzzle:
            for value in row:
                print(value, end=' ')
            print()

    def next_unassigned(self, index):
        # looks for the next unassigned spot
        for i in range(9):
            for j in range(9):
                if self.puzzle[i][j] == 0:
                    # index get the position of the unassigned spot
                    index[0] = i
                    index[1] = j
                    return True
        return False

    def solve_puzzle(self):
        index = [0, 0]
        if not self.next_unassigned(index):
            # if all spots are assigned then we are complete
            return True

        # values were obtained from next_unassigned
        x = index[0]
        y = index[1]

        # array of all possible values
        possible_values = [x for x in range(1, 10)]

        # removing values that are already used
        self.remove_row_values(possible_values, x)
        self.remove_column_values(possible_values, y)
        self.remove_box_values(possible_values, x, y)

        for value in possible_values:
            # attempt value in unassigned spot
            self.puzzle[x][y] = value

            # calls func again until all spots are assigned properly
            if self.solve_puzzle():
                return True

            # attempted value breaks future assignments, assign spot to empty value
            self.puzzle[x][y] = 0

        # list of possible values does not follow rule, previous assignment will be undone
        return False


if __name__ == "__main__":
    solve = Sudoku('puzzle.txt')
    if solve.solve_puzzle():
        solve.print_grid()
    else:
        print("No solution")

File: Sudoku/Sudoku.java
package test;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

public class Sudoku {
	char[][] grid;

	public Sudoku(String filename) throws IOException {
		// setup sudoku puzzle
		this.read_file(filename);
	}

	public void read_file(String file_name) throws IOException {
		File f = new File(file_name);
		BufferedReader br = new BufferedReader(new FileReader(f));
		String row_input;
		this.grid = new char[9][9];

		for(int row_num = 0; (row_input= br.readLine()) != null; row_num++) {
		    String values[] = row_input.split(" ");
		    char[] row = new char[9];

		    for(int i = 0; i < values.length; i++)
		    	row[i] = values[i].charAt(0);

		    this.grid[row_num] = row;
		}

		br.close();
	}

	public void print_puzzle() {
		for(int i = 0; i < 9; i++) {
			for(int j = 0; j < 9; j++) {
				System.out.print(this.grid[i][j]);
				System.out.print(' ');
			}
			System.out.println();
		}}

	public boolean search_for_empty(int[] index) {
		// changes index and returns true if there is another unassigned spot else return false
		for(int i = 0; i < 9; i++)
			for(int j = 0; j < 9; j++)
				if(this.grid[i][j] == '0') {
					index[0] = i;
					index[1] = j;
					return true;
				}
		return false;
	}

	public char[] remove_duplicates(char[] l, int x, String version){
		// depending on version, will remove numbers that already used in row or column
		for(int i = 0; i < 9; i++) {
			char num = version.equals("rows") ? this.grid[x][i] : this.grid[i][x];
			if(this.in_array(l, num))
				l = this.remove_from_array(l, num);
		}
		return l;
	}

	public char[] remove_duplicates_from_square(char[] l, int x, int y){
		// start and end for 3x3 box
		int[] x_range = this.box_range(x);
		int[] y_range = this.box_range(y);

		// will remove numbers that already used in 3x3
		for(int i = x_range[0]; i < x_range[1]; i++)
			for(int j = y_range[0]; j < y_range[1]; j++) {
				char num = this.grid[i][j];
				if(this.in_array(l, num))
					l = this.remove_from_array(l, num);
			}
		return l;
	}

	public int[] box_range(int num) {
		int[] range = new int[2];
        for(int i = 0; i < 4; i++)
            if(i * 3 <= num && num < (i + 1) * 3) {
            	range[0] = i * 3;
            	range[1] = (i + 1) * 3;
            }
        return range;
	}

	public boolean in_array(char[] l, char value) {
		// return true if array contains value else return false
		for(int i = 0; i < l.length; i++)
			if(l[i] == value)
				return true;
		return false;
	}

	public char[] remove_from_array(char[] l, char value) {
		char[] removed = new char[l.length - 1];
		// removes value from array
		for(int i = 0; i < l.length; i++)
			if(l[i] == value) {
				 System.arraycopy(l, 0, removed, 0, i);
				 System.arraycopy(l, i+1, removed, i, l.length - i - 1);
			}
		return removed;
	}

	public boolean solve_sudoku() {
		int[] index = {0, 0};

		// looks for empty spot and gives coordinates in index
		// if there are no more spots then we are finished
		if(!this.search_for_empty(index))
			return true;

		int x = index[0];
		int y = index[1];

		char[] possible_values = new char[9];
		for(int i = 1; i < 10; i++)
			possible_values[i - 1] = (char) (i + '0');

		// remove values are used in specific row, column, and 3x3 box
		possible_values = this.remove_duplicates(possible_values, x, "rows");
		possible_values = this.remove_duplicates(possible_values, y, "columns");
		possible_values = this.remove_duplicates_from_square(possible_values, x, y);

		for(int i = 0; i < possible_values.length; i++) {
			// assign spot with a valid input
			this.grid[x][y] = possible_values[i];

			//continues until all spots have valid assignments
			if(this.solve_sudoku())
				return true;

			// assigned input is not valid so we reset
			this.grid[x][y] = '0';
		}
		return false;
	}

	public static void main(String[] args) {
		try {
			String filename = "path_to\\puzzle.txt";
			Sudoku sudoku = new Sudoku(filename);
			if(sudoku.solve_sudoku())
				sudoku.print_puzzle();
			else
				System.out.print("no solution");
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}
}
